### PR TITLE
loading: provide code for reifying a load-path lookup result

### PR DESCRIPTION
**Pull request by vtjnash**
This allows asking (and storing) queries about what total set of modules might need to be loaded, before starting any load work. This is base work for later working on https://github.com/JuliaLang/julia/issues/28781, so we might add additional information later (such as folding `concrete_deps` into it?). This is intended to allow us to pass around the dependency list explicitly, so that worker processes (such as during module loading, precompile, Distributed, and so on) can be isolated from the implicit and mutable LOAD_PATH and Project file parts of the system, once a load is triggered by a toplevel `using` or `cachecompile` call.

### CODE REVIEW

File: base/loading.jl [hunk 1]
```diff
@@ -227,7 +251,7 @@ function locate_package(pkg::PkgId, cache::TOMLCache=TOMLCache())::Union{Nothing
                 return implicit_manifest_uuid_path(env, pkg, cache)
             end
             @assert found.uuid !== nothing
-            return locate_package(found, cache) # restart search now that we know the uuid for pkg
+            return locate_package(found, cache) # restart search now that we know the uuid for pkg (TODO: the existence of this line of code is probably a bug)
```
Comment (KristofferC) [inline base/loading.jl:254]: Yeah, this line feels strange to me too.

File: base/loading.jl [hunk 2]
```diff
@@ -371,6 +425,7 @@ end
 
 # given a directory (implicit env from LOAD_PATH) and a name,
 # check if it is an implicit package
+# TODO: aren't we supposed to first check for the Project file first and see if it declares a path?
```
Comment (KristofferC) [inline base/loading.jl:428]: the path entry in `Project` the project file seems to me like a bit of an experiment that turned out to be unused. Pkg doesn't know about it for example.

File: base/loading.jl [hunk 3]
```diff
@@ -459,18 +541,54 @@ function explicit_manifest_deps_get(project_file::String, where::UUID, name::Str
...
+    found_where = false
+    entries = get(d, where.name, nothing)::Union{Vector{Any}, Nothing}
+    if entries !== nothing
+        for entry in entries
+            entry::Dict{String, Any}
+            uuid = get(entry, "uuid", nothing)::Union{String, Nothing}
+            uuid === nothing && continue
+            if UUID(uuid) === where.uuid
+                found_where = true
+                # deps is either a list of names (deps = ["DepA", "DepB"]) or
```
Comment (KristofferC) [inline base/loading.jl:561]: Could some of this code be shared with `explicit_manifest_deps_get`?

**Review by KristofferC — COMMENTED** [review]
This does indeed seem useful to have!

**Comment by KristofferC** [comment]
> This is intended to allow us to pass around the dependency list explicitly, so that worker processes (such as during module loading, precompile, Distributed, and so on) can be isolated from the implicit and mutable LOAD_PATH and Project file parts of the system, once a load is triggered by a toplevel using or cachecompile call.

In what way did you imagine passing this information to worker. `repr` feels a bit too slow when large environment data has to be passed (#37989), and we don't really have access to serialization in Base.

**Comment by vtjnash** [comment]
This doesn’t need either form. It’s just a trivial key/value store with a tiny, well-defined set of pairs. We actually already deal with these, just only sending a large subset of them instead of all.

**Comment by KristofferC** [comment]
Rebased

**Comment by vtjnash** [comment]
@KristofferC is this PR still of any interest?

**Pull request closed without merging**